### PR TITLE
[UT coverage] improve the mul_mkldnn_op line coverage

### PR DESCRIPTION
**Pull request by lidanqing-vv**
Locally, mul_mkldnn_op.cc line coverage is 97%

### CODE REVIEW

**Comment by bingyanghuang** [comment]
@lidanqing-intel Please don't forget to add `test=develop` in your each commits.

**Comment by lidanqing-vv** [comment]
> @lidanqing-intel Please don't forget to add `test=develop` in your each commits.

I am still fixing, fp32 mul was not tested before and now one CAPI failed.

**Comment by lidanqing-vv** [comment]
@luotao1 Since 29th Jan, Paddle_PR_CI_Coverage_Night crashed, We can not see most recent mkldnn line coverage. (There are 2 new UT merged). Will somebody fix the nightly machine?

**Comment by luotao1** [comment]
@tianshuo78520a will help see it.

**Comment by tianshuo78520a** [comment]
I debugged it today, and tonight, I expect the task to pass normally tomorrow.

**Comment by tianshuo78520a** [comment]
New  Paddle_PR_CI_Coverage_Night is Success [Link](http://ci.paddlepaddle.org/viewLog.html?buildId=276831&buildTypeId=Paddle_PrCiCoverageNight&tab=report_project7_Python_Full_Coverage)

**Comment by lidanqing-vv** [comment]
Hi, @wojtuss Please review this PR

File: paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc [hunk 1]
```diff
@@ -60,17 +67,120 @@ class MulPrimitiveFactory {
...
+      auto scale_y = ctx.Attr<std::vector<float>>("scale_y");
+      y_input_ = QuantInputY(trans_y, scale_y);
+    } else {
+      y_input_ = TransposeInputY(&y_matrix);
+    }
+
     auto dst_desc = CreateMemDescriptor<OT>(output, MKLDNNMemoryFormat::any);
 
-    mul_ = CreateMulPrimitive(*x_input_, *y_input_, dst_desc, output, ctx);
+    mul_ = CreateMemMulPrimitive(*(x_input_), *(y_input_), dst_desc, output,
```
Comment (wojtuss) [inline paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc]: The parentheses around the pointers can be skipped.

File: paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc [hunk 2]
```diff
@@ -60,17 +67,120 @@ class MulPrimitiveFactory {
...
+        output_shift_scale[i] =
+            scale_out_data / (scale_x_data * scale_y_data[i]);
+    }
+    int mul_mask = is_multi_channel ? 1 : 0;
+    mul_attr.set_output_scales(mul_mask, output_shift_scale);
+
+    return mul_attr;
+  }
+
+  inner_product_forward CreateMemMulPrimitive(const memory &x_memory,
```
Comment (wojtuss) [inline paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc]: The `Mem` in the name can be misleading, as we do not create a memory (primitive), just an op primitive. Although the name `CreateMulPrimitive` is overloaded, I would keep it.

File: paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc [hunk 3]
```diff
@@ -60,17 +67,120 @@ class MulPrimitiveFactory {
...
+    } else {
+      mul_prim_desc = inner_product_forward::primitive_desc(mul_desc, engine_);
+    }
+
+    output_ = CreateDstMemory(mul_prim_desc, ctx, output);
+
+    return inner_product_forward(mul_prim_desc);
+  }
+
+  inner_product_forward::primitive_desc CreateMulPrimDesc(
```
Comment (wojtuss) [inline paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc]: This method is unused

File: paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc [hunk 4]
```diff
@@ -81,7 +191,6 @@ class MulPrimitiveFactory {
     astream.wait();
   }
 
- protected:
```
Comment (wojtuss) [inline paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc:84]: The below stuff can be still kept private.

File: paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc [hunk 5]
```diff
@@ -379,19 +321,19 @@ inner_product_forward GetMulPrimitive(const MKLDNNDeviceContext &dev_ctx,
...
+  constexpr bool is_int8 =
       std::is_same<XT, int8_t>::value || std::is_same<XT, uint8_t>::value;
   bool force_fp32_output = ctx.Attr<bool>("force_fp32_output");
 
-  if (enable_quant && !force_fp32_output) {
+  if (is_int8 && !force_fp32_output) {
     return GetPrimitiveFactory<XT, YT, int8_t>(dev_ctx, ctx, input_x, input_y,
-                                               mkldnn_engine, enable_quant)
-        ->CreateMulPrimitive(input_x, input_y, output, ctx);
+                                               mkldnn_engine, is_int8)
```
Comment (wojtuss) [inline paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc]: I think adding `is_int8` as a `static constexpr bool` field in the `MulPrimitiveFactory` would save us passing the `is_int8` argument to a few methods.

File: python/paddle/fluid/tests/unittests/mkldnn/test_mul_int8_mkldnn_op.py [hunk 6]
```diff
@@ -52,11 +52,11 @@ def init_data(self):
 
         # limit random range inside |-127, 127| to avoid overflow on SKL
         if self.srctype == np.int8:
-            A_data = np.random.randint(-127, 127, (2, 5)).astype(np.int8)
-        else:
-            A_data = np.random.randint(0, 127, (2, 5)).astype(np.uint8)
+            A_data = np.random.randint(-127, 127, (20, 5)).astype(np.int8)
+        elif self.srctype == np.uint8:
```
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_mul_int8_mkldnn_op.py]: Is there an option other than `np.int8` and `np.uint8`? If yes, please add `else`. If not, please keep `else` here.

**Comment by lidanqing-vv** [comment]
Hi, @wojtuss I updated according to reviews, please take a look.

**Review by wojtuss — APPROVED** [review]
LGTM

**Pull request merged**